apps/src/studio/StudioAnimation.js

```diff
@@ -176,32 +241,65 @@ StudioAnimation.prototype.redrawCenteredAt = function (center) {
  * Sets which animation to play out of the sprite sheet.
  * Animations are indexed by their position in the sprite sheet, where each
  * animation is its own column and animation zero is the far-left column.
+ * @param {!string} animationType
  * @param {!number} animationIndex
  */
-StudioAnimation.prototype.setCurrentAnimation = function (animationIndex) {
-  this.currentAnimation_ = animationIndex;
+StudioAnimation.prototype.setCurrentAnimation = function (animationType, animationIndex) {
+  this.currentAnimationType_ = animationType;
+  this.currentAnimationIndex_ = animationIndex;
 };
 
 /**
- * Set the animation rate for this item's sprite.
- * @param {number} framesPerSecond
+ * Creates a new special animation types based on specific frames to play from
+ * the sprite sheet.
+ * @param {!string} type - the name of the new animation type
+ * @param {!number} index - the index of the new animation
+ * @param {!array} animationList - an array with frame information
+ * @param {string} [animationList[].type] - animation type for a specific frame.
+ * @param {number} [animationList[].index] - animation index for a specific frame.
+ * @param {number} [animationList[].frame] - animation frame for a specific frame.
  */
-StudioAnimation.prototype.setAnimationRate = function (framesPerSecond) {
-  if (this.animator_) {
-    window.clearInterval(this.animator_);
+StudioAnimation.prototype.createSpecialAnimation =
+    function (type, index, animationList) {
+  if (!this.specialAnimations_[type]) {
+    this.specialAnimations_[type] = [];
+  }
+  var frames = [];
+  for (var i = 0; i < animationList.length; i++) {
+    frames.push(
+        this.spriteSheet_.getFrame(animationList[i].type,
+            animationList[i].index,
+            animationList[i].frame));
   }
-  this.animator_ = window.setInterval(function () {
-    if (this.loop_ || this.currentFrame_ + 1 < this.spriteSheet_.framesPerAnimation) {
-      this.currentFrame_ = (this.currentFrame_ + 1) %
-          this.spriteSheet_.framesPerAnimation;
-    }
-  }.bind(this), Math.round(1000 / framesPerSecond));
+  this.specialAnimations_[type][index] = frames;
 };
 
 /**
- * Change visible opacity of this animation..
+ * Set the animation speed for this item's sprite.
+ * @param {number} ticksPerFrame
+ */
+StudioAnimation.prototype.setAnimationFrameDuration = function (ticksPerFrame) {
+  this.animationFrameDuration_ = ticksPerFrame;
+};
+
+/**
+ * Change visible opacity of this animation.
  * @param {number} newOpacity (between 0 and 1)
  */
 StudioAnimation.prototype.setOpacity = function (newOpacity) {
   this.opacity_ = newOpacity;
 };
+
+/**
+ * Make this animation hidden.
+ */
+StudioAnimation.prototype.hide = function () {
+  this.element_.setAttribute('visibility', 'hidden');
+};
+
+/**
+ * Make this animation visible.
+ */
+StudioAnimation.prototype.show = function () {
+  this.element_.setAttribute('visibility', 'visible');
```

Comment:
Probably out of scope for what's already such a big change, but I wonder if we could start moving towards a slightly more react-style "immediate mode" renderer where operations such as these just set a flag, and then the render function can update the appearance of the element in a more centralized location every frame.  (And it can cache its known render state to improve performance if needed.)